.github/workflows/validate.yml

```diff
@@ -77,22 +77,9 @@ jobs:
     runs-on: ubuntu-20.04
     steps:
     - uses: actions/checkout@v3
-    - name: vars
-      run: |
-        echo "VERSION=3.3.1" >> $GITHUB_ENV
-        echo "$(go env GOPATH)/bin" >> $GITHUB_PATH
-    - name: cache go mod and $GOCACHE
-      uses: actions/cache@v2
+    - uses:  mfinelli/setup-shfmt@v1
```

Comment:
I'm slightly on the fence on running a random action in CI;  seems like the "old" code was more complicated than it had to be? Would just a one-line `go install` work?;

```
GO111MODULE=on go inastall mvdan.cc/sh/v3/cmd/shfmt@v3.3.1
```

Reply:
The one-liner won't work because at the very least we need to add `GOPATH/bin` to `$PATH` (which the old code did).

For the newer shfmt, I guess, it won't work as ubuntu-20.04 used in this job comes with go 1.13.8, which might not be enough for shfmt 3.4.3.

Using a pre-released binary is better because:
 - no need to install golang
 - no need to download dependencies and build
 - the released binary is more stable/predictable than the one we build ourselves

If you don't like using a third-party action (I took a look at it and it seems short, focused, and being maintained), I can certainly rewrite this using wget (since it's Linux only we don't have to care about .exe so it should be simple enough).

Let me know how do you want this to be handled.

Reply:
It looks like Go 1.15 and up are on those runners (from the link in the workflow); https://github.com/actions/virtual-environments/blob/ubuntu20/20220227.1/images/linux/Ubuntu2004-Readme.md#go

Well, they're listed under "cached tools", so I _think_ they're there? (but with a custom path, so perhaps it expects the env-var to be se with go-setup); of course there's always the docker image we could run (and just 1 mb); https://hub.docker.com/r/mvdan/shfmt

Reply:
@thaJeztah PTAL ^^

Reply:
@thaJeztah PTAL ^^

Reply:
Fetching the prebuilt binary sounds good to me.

Reply:
If we stick with the action, at least we should pin it to a specific version.

I'm good with downloading the binary; that said; the docker image _is a single-binary image_, and docker should be present in GHA so would make it easier to run (and it's multi-arch, so no need to take different architectures into account)

Slightly orthogonal, but looking at the make target we currently use, it looks like we don't need to use two separate runs for `bats` and `bash` files; `shfmt` automatically detects what to use based on the file-extension, and once https://github.com/mvdan/sh/commit/a876f44daf649bc0184b0ad7685224a65fbce8f0 is in a release, even would detect a shebang in the file. Given that our `bats` files have a `.bats` extension, it looks like we don't need the shebang detection, so (using the docker image, but same for binary) so it looks like we can combine both runs (omitting the `-ln` option);

```bash
docker run --rm -v $(pwd):/src -w /src mvdan/shfmt -d -w man/*.sh script/* tests/*.sh tests/integration/*.bash tests/integration/*.bats
```

Reply:
> it looks like we don't need to use two separate runs for `bats` and `bash` files

Yes, I do have that in mind and even filed a few issues to shfmt for that to happen. Looks like it's fully working now so I added a commit that does that.

I don't like using docker from `Makefile` for shfmt; the main reason is I can use custom `shfmt` binary to check things out (and I did a number of times when helping in its development and testing).

Using docker from CI job file is all right I guess; the only problem is we have to repeat the `tests/integration/*.bats man/*.sh script/* tests/*.sh tests/integration/*.bash` line again, which creates a ever-so-slight maintenance burden (and a possiblity that CI job and the makefile target will stop being in sync).

A solution for that would be using `shfmt -f .`. I remember it was not working before for some reason (probably because we needed separate lists for bash and bats). Actually, it works now, so we no longer have to supply the fixed list of paths.

Switching this PR to use docker.

> If we stick with the action, at least we should pin it to a specific version.

Shouldn't we do that with docker image, too?

Reply:
Damn. shfmt 3.4.3 can not yet mix bats and bash in the same invocation (i.e. `-ln bash` and `-ln bats` are still required). So everything I did with Docker does not work.

Setting this on pause until shfmt 3.5.0 is released.